src/osd/ReplicatedPG.cc

```diff
@@ -3285,14 +3285,12 @@ int ReplicatedPG::do_tmapup(OpContext *ctx, bufferlist::iterator& bp, OSDOp& osd
   return result;
 }
 
+//EFBIG if attempt to WRITE/WRITEFULL/TRUNCATE beyond max, and length 0 writes are allowed.
 static int check_offset_and_length(uint64_t offset, uint64_t length, uint64_t max)
 {
-  if (offset >= max ||
-      length > max ||
-      offset + length > max)
-    return -EFBIG;
-
-  return 0;
+  if (offset + length < max || (offset + length == max && length) )
```

Comment:
The reason to do the test the other way is because an extremely large offset and/or length could overflow the integer so offset + length becomes small.

The way the test is written now isn't perfect either with very large value of max where offset and length are reasonable but offset + length overflows the integer.